src/detectors/BTOF/BarrelTOFNeighborFinder.cc

```diff
@@ -0,0 +1,201 @@
+  _initWithCell(hitCell);
+  // look for cache. If exist, return cache value
+  if (_useCache) {
+    auto it = _cache.find(hitCell);
+    if (it != _cache.end())
+      return it->second;
+  }
+
+  std::unordered_set<dd4hep::rec::CellID> dp;
+  // fill cache
+  std::shared_ptr<std::vector<dd4hep::rec::CellID>> neighbors =
+      std::make_shared<std::vector<dd4hep::rec::CellID>>();
+  this->_findAllNeighborsInSensor(hitCell, neighbors, dp);
+
+  if (_useCache)
+    for (auto cell : *neighbors)
+      _cache[cell] = neighbors;
+
+  return neighbors;
+}
+
+const dd4hep::rec::CellID
+BarrelTOFNeighborFinder::_getSensorID(const dd4hep::rec::CellID& hitCell) {
+  _initWithCell(hitCell);
+  // fix x-y, what you left with are ids that corresponds to sensor info
+  // cellID may change when position changes.
+  auto sensorID = hitCell; //_converter -> cellID(_converter -> position(hitCell));
+  _decoder->set(sensorID, "x", 0);
+  _decoder->set(sensorID, "y", 0);
+
+  return sensorID;
+}
+
+/****************************************
+ * Functions below aren't needed anymore to find neighbors
+ * They are still handy so I keep them
+ * *************************************/
+dd4hep::Position BarrelTOFNeighborFinder::cell2GlobalPosition(const dd4hep::rec::CellID& cell) {
+  if (!_converter)
+    throw std::runtime_error(
+        "CellIDPositionConverter not initialized in BarrelTOFNeighborFinder. Have you called "
+        "BarrelTOFNeighborFinder::init(/* detector */) first?");
+  return _converter->position(cell);
+}
+
+dd4hep::rec::CellID BarrelTOFNeighborFinder::globalPosition2Cell(const dd4hep::Position& pos) {
+  if (!_converter)
+    throw std::runtime_error(
+        "CellIDPositionConverter not initialized in BarrelTOFNeighborFinder. Have you called "
+        "BarrelTOFNeighborFinder::init(/* detector */) first?");
+  return _converter->cellID(pos);
+}
+
+dd4hep::Position
+BarrelTOFNeighborFinder::local2GlobalInStaveFromCell(const dd4hep::rec::CellID& cell,
+                                                     const dd4hep::Position& pos) {
+  // convert local position to global position
+  // assuming the local position is located at the same volume as cell
+  if (!_detector)
+    throw std::runtime_error("Detector ptr not initialized in BarrelTOFNeighborFinder. Have you "
+                             "called BarrelTOFNeighborFinder::init(/* detector */) first?");
+
+  auto geoManager = _detector->world().volume()->GetGeoManager();
+  auto position   = this->cell2GlobalPosition(cell);
+  auto node       = geoManager->FindNode(position.x(), position.y(), position.z());
+  auto currMatrix = geoManager->GetCurrentMatrix();
+
+  double g[3], l[3];
+  pos.GetCoordinates(l);
+  currMatrix->LocalToMaster(l, g); // <- only different from cell2LocalPosition
+  position.SetCoordinates(g);      // <- only different from cell2LocalPosition
+  return position;
+}
+
+dd4hep::Position BarrelTOFNeighborFinder::global2Local(const dd4hep::Position& pos) {
```

Comment:
This is always a frustrating function to do from the dd4hep geometry but I think it can be simplified a bit, will have to check.

Reply:
Turns out "global2Local" is not required for digitization, but I do need "cell2LocalPosition" so I can integrate the charge deposition inside a cell in sensor frame. Please see detector/BTOFChargeSharing

Reply:
I think you did want the global2Local function as it used the actual position of the geant4 hit as the centroid rather than the position of the cell.

Reply:
Oh you made a fantastic point. I do need that. Mistake is made where I just fetch the center of the cell in charge sharing. Let me put it back in.